R/logging.R

```diff
@@ -63,6 +63,157 @@ message_wrap <-
+# `initialize_catalog()` inside of machinery that is wrapped by other tuning
+# method. e.g. `tune_bayes()` has an issue cataloger that ought not to be
+# overwritten when fitting over an initial grid.
+catalog_is_active <- function() {
+  tune_env$progress_active
+}
+
+# initializes machinery for the tune catalog inside of an environment.
+# the `env` should be an execution environment that persists throughout the
+# tuning process for a given tuning approach and exits once tuning is completed.
+#' @rdname tune-internal-functions
+#' @export
+initialize_catalog <- function(env = rlang::caller_env(), control) {
+  catalog <-
+    tibble::tibble(
+      type = character(0),
+      note = character(0),
+      n = numeric(0),
+      id = numeric(0)
+    )
+
+  if (!(allow_parallelism(control$allow_par) || is_testing())) {
+    progress_active <- TRUE
+  } else {
+    progress_active <- FALSE
+  }
+
+  rlang::env_bind(tune_env, progress_env = env)
+
+  rlang::env_bind(tune_env, progress_catalog = catalog)
+  withr::defer(
+    rlang::env_bind(tune_env, progress_catalog = NULL),
+    envir = env
+  )
+
+  rlang::env_bind(tune_env, progress_active = progress_active)
+  withr::defer(
+    rlang::env_bind(tune_env, progress_active = FALSE),
+    envir = env
+  )
+
+
+  invisible(NULL)
+}
+
+# given a catalog, summarize errors and warnings in a 1-length glue vector.
+# for use by the progress bar inside of `tune_catalog()`.
+summarize_catalog <- function(catalog) {
+  if (nrow(catalog) == 0) {
+    return("")
+  }
+
+  res <- dplyr::arrange(catalog, id)
+  res <- dplyr::mutate(res, color = dplyr::if_else(type == "warning", list(cli::col_yellow), list(cli::col_red)))
+  res <- dplyr::rowwise(res)
+  res <- dplyr::mutate(res, msg = glue::glue("{color(cli::style_bold(id))}: x{n}"))
+  res <- dplyr::ungroup(res)
+  res <- dplyr::pull(res, msg)
+  res <- glue::glue_collapse(res, sep = "   ")
+
+  res
+}
+
+# a light wrapper around `tune_catalog()` for use inside of `tune_log()`
+log_catalog <- function(msg, type) {
+  type <-
+    switch(
+      type,
+      warning = "warning",
+      danger = "error",
+      return(invisible(NULL))
+    )
+
+  issues <-
+    tibble::tibble(
+      type = type,
+      note = msg
+    )
+
+  tune_catalog(issues)
+
+  invisible(NULL)
+}
+
+# an alternative to `tune_log()` that maintains a "catalog" of previously
+# encountered issues, and interactively summarizes them by type rather than
+# printing out each new tuning issue individually.
+tune_catalog <- function(issues) {
+  catalog <- rlang::env_get(env = tune_env, nm = "progress_catalog")
+
+  res <- dplyr::count(issues, type, note) %>% mutate(id = NA_integer_)
+  res <- dplyr::bind_rows(res, catalog)
+  res <- dplyr::group_by(res, type, note)
+  # dplyr::first will gain an `na_rm` argument in 1.1.0
+  res <- dplyr::summarize(res, n = sum(n), id = dplyr::first(id[!is.na(id)]))
+  res <- dplyr::ungroup(res)
```

Comment:
Do you want to leave it grouped by `type` and `note`?

Reply:
I don't believe so. I don't think we rely on grouping / re`group_by()` later on, so I figure removing groups is the more cautious thing to do.

Reply:
Oh I was totally just reading this wrong. Completely ignore this, I think what you have is right

Reply:
Ah, okay! Sounds good.